database/plugin/metadata/sqlite/transaction.go

```diff
@@ -51,40 +58,177 @@ func (d *MetadataStoreSqlite) SetTransaction(
+	for _, input := range tx.Inputs() {
+		inTxId := input.Id().Bytes()
+		inIdx := input.Index()
+		utxo, err := d.GetUtxo(inTxId, inIdx, txn)
+		if err != nil {
+			return fmt.Errorf(
+				"failed to fetch input %x#%d: %w",
+				inTxId,
+				inIdx,
+				err,
+			)
+		}
+		if utxo == nil {
+			// Do nothing
+			continue
+		}
+		tmpTx.Inputs = append(
+			tmpTx.Inputs,
+			*utxo,
+		)
+	}
+	// Add Collateral to Transaction
+	if len(tx.Collateral()) > 0 {
+		var caseClauses []string
+		var whereConditions []string
+		var caseArgs []any
+		var whereArgs []any
+
+		for _, input := range tx.Collateral() {
+			inTxId := input.Id().Bytes()
+			inIdx := input.Index()
+			utxo, err := d.GetUtxo(inTxId, inIdx, txn)
+			if err != nil {
+				return fmt.Errorf(
+					"failed to fetch input %x#%d: %w",
+					inTxId,
+					inIdx,
+					err,
+				)
+			}
+			if utxo == nil {
+				// Do nothing
+				continue
+			}
+			caseClauses = append(caseClauses, "WHEN tx_id = ? AND output_idx = ? THEN ?")
+			caseArgs = append(caseArgs, inTxId, inIdx, txHash)
+			whereConditions = append(whereConditions, "(tx_id = ? AND output_idx = ?)")
+			whereArgs = append(whereArgs, inTxId, inIdx)
+			tmpTx.Collateral = append(
+				tmpTx.Collateral,
+				*utxo,
+			)
+		}
+		if len(caseClauses) > 0 {
+			args := append(caseArgs, whereArgs...)
+			sql := fmt.Sprintf(
+				"UPDATE utxos SET collateral_by_tx_id = CASE %s ELSE collateral_by_tx_id END WHERE %s",
+				strings.Join(caseClauses, " "),
+				strings.Join(whereConditions, " OR "),
+			)
+			result = txn.Exec(sql, args...)
+			if result.Error != nil {
+				return fmt.Errorf("batch update collateral: %w", result.Error)
+			}
+		}
+	}
+	// Add ReferenceInputs to Transaction
+	if len(tx.ReferenceInputs()) > 0 {
+		var caseClauses []string
+		var whereConditions []string
+		var caseArgs []any
+		var whereArgs []any
+
+		for _, input := range tx.ReferenceInputs() {
+			inTxId := input.Id().Bytes()
+			inIdx := input.Index()
+			utxo, err := d.GetUtxo(inTxId, inIdx, txn)
+			if err != nil {
+				return fmt.Errorf(
+					"failed to fetch input %x#%d: %w",
+					inTxId,
+					inIdx,
+					err,
+				)
+			}
+			if utxo == nil {
+				// Do nothing
+				continue
+			}
+			caseClauses = append(caseClauses, "WHEN tx_id = ? AND output_idx = ? THEN ?")
+			caseArgs = append(caseArgs, inTxId, inIdx, txHash)
+			whereConditions = append(whereConditions, "(tx_id = ? AND output_idx = ?)")
+			whereArgs = append(whereArgs, inTxId, inIdx)
+			tmpTx.ReferenceInputs = append(
+				tmpTx.ReferenceInputs,
+				*utxo,
+			)
+		}
+		if len(caseClauses) > 0 {
+			args := append(caseArgs, whereArgs...)
+			sql := fmt.Sprintf(
+				"UPDATE utxos SET referenced_by_tx_id = CASE %s ELSE referenced_by_tx_id END WHERE %s",
+				strings.Join(caseClauses, " "),
+				strings.Join(whereConditions, " OR "),
+			)
+			result = txn.Exec(sql, args...)
+			if result.Error != nil {
+				return fmt.Errorf("batch update reference inputs: %w", result.Error)
+			}
+		}
+	}
```

Comment:
Some inline comments would be useful for describing what's going on with this query building